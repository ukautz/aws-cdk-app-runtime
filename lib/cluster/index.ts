export * from './cluster';
export * from './router';
export { specKeys as clusterSpecKeys, Specs as ClusterSpecs } from './specs';
