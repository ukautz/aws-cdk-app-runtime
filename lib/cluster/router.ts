import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as iam from '@aws-cdk/aws-iam';
import * as route53 from '@aws-cdk/aws-route53';
import * as targets from '@aws-cdk/aws-route53-targets';
import * as servicediscovery from '@aws-cdk/aws-servicediscovery';
import * as service from '../service';
import * as util from '../util';
import { loadCertificate } from '../util/loaders';

const defaultEnforceHttps = true;
const defaultRouterImage = 'traefik';
const defaultRouterImageVersion = '2.4';

export interface RouterProps extends RouterPropsExposed {
  readonly cluster: ecs.ICluster;
  readonly certificate?: boolean | string | acm.ICertificate;
  readonly hostedZone: route53.IHostedZone;
  readonly publicDomain: string;
}
export interface RouterPropsExposed {
  /**
   * Whether to enforce HTTPs (redirect all incoming HTTP requests to HTTPS)
   */
  readonly enforceHttps?: boolean;

  /**
   * Optional traffic route condition in the load balancer (e.g. when behind CloudFront to assure only CF sends traffic)
   */
  readonly requireHeader?: { name: string; values: string[] };

  /**
   * Optionally provided logging settings. Either a proper log driver, or false
   * to disable logging or a string used a as prefix for CloudWatch logs or
   * default value for cloudwatch logs
   */
  readonly logging?: ecs.LogDriver | string | false;

  /**
   * The log level for Traefik
   *
   * @see https://doc.traefik.io/traefik/observability/logs/#level
   * @default ERROR
   */
  readonly logLevel?: string;

  /**
   * The resources that the router service should have
   */
  readonly resources?: string | util.ResourcesProps;

  /**
   * The image the container is to be started from. Can be:
   * * ecr:<arn>
   * * path:</to/folder/with/dockerfile>
   * * <dockerhub image>
   *
   * @see https://hub.docker.com/_/traefik
   * @default traefik (dockerhub image)
   */
  readonly image?: string | ecs.ContainerImage;

  /**
   * The tag / version of the image to be run.
   * * for ecr:
   *
   * @see https://hub.docker.com/_/traefik
   * @default 2.4
   */
  readonly imageVersion?: string;
}

export interface RouterSpecs {
  /**
   * ARN of the (fargate) ECS <TODO> that
   */
  routerArn: string;

  /**
   * ARN of the certificate that terminates the domain and wildcard.domain
   */
  routerCertificateArn?: string;

  /**
   * ARN of the load balancer in-front of the router service
   */
  routerLoadBalancerArn: string;

  /**
   * DNS name of the load balancer in-front of the router service
   */
  routerLoadBalancerDnsName: string;

  /**
   * ID of the security group associated with the load balancer
   */
  routerLoadBalancerSecurityGroupId: string;

  /**
   * ARN of the IAM role that the router service assumes at runtime
   */
  routerRoleArn: string;

  /**
   * Security group IDs of the router service, separated by ","
   */
  routerSecurityGroupIds: string;

  /**
   * ARN of the TaskDefinition that implements the router service
   */
  routerTaskDefinitionArn: string;
}

/**
 * Router takes exposes public services under a shared public domain. http(s)://service-name.public.domain/
 *
 * Components:
 * - A container image of Traefik
 * - An ECS Fargate task that runs the container image and exposes it via
 * - An ALB, offering HTTP and HTTPS, using a provided certificate (optional HTTP -> HTTPS redirect)
 * - A certificate for the wildcard of the (cluster) public domain (`*.public.domain`) securing the ALB
 * - A wildcard record in the cluster hosted zone pointing to the ALB
 *
 * Integration:
 * - IGrantable: So that running ECS service can access AWS API (e.g. Traefik scans ECS for services it should route to)
 * - IConnectable: So that running ECS service can access
 *
 * Considerations:
 * - Multiple routing models? This is a shared model, using a hostname based routing with Traefik behind an ALB. Other models could dedicate resources (i.e. an ALB per service) etc.
 */
export class Router extends cdk.Construct implements iam.IGrantable, ec2.IConnectable {
  public readonly service: ecs.FargateService;
  public readonly loadBalancer: elbv2.IApplicationLoadBalancer;
  public readonly grantPrincipal: iam.Role;
  public readonly certificate?: acm.ICertificate;
  private readonly resources: util.Resources;

  constructor(scope: cdk.Construct, id: string, props: RouterProps) {
    super(scope, id);

    this.resources =
      typeof props.resources === 'string'
        ? util.Resources.fromString(props.resources)
        : new util.Resources(props.resources);

    this.certificate = this.initCertificate(props);
    this.grantPrincipal = this.createServiceTaskRole(props);

    const image = this.initImage(props);
    this.service = this.createService(props, image);
    this.loadBalancer = this.initLoadBalancer(props);
    this.setupAutoScaling();
    this.routeToLoadBalancer(props);
    this.publishInClusterNamespace(props);
  }

  public get connections(): ec2.Connections {
    return this.service.connections;
  }

  public get specs(): RouterSpecs {
    return {
      routerArn: this.service.serviceArn,
      routerCertificateArn: this.certificate?.certificateArn,
      routerLoadBalancerArn: this.loadBalancer.loadBalancerArn,
      routerLoadBalancerDnsName: this.loadBalancer.loadBalancerDnsName,
      routerLoadBalancerSecurityGroupId: this.loadBalancer.connections.securityGroups
        .map((sg) => sg.securityGroupId)
        .join(','),
      routerRoleArn: this.service.taskDefinition.taskRole.roleArn,
      routerSecurityGroupIds: this.service.connections.securityGroups.map((sg) => sg.securityGroupId).join(','),
      routerTaskDefinitionArn: this.service.taskDefinition.taskDefinitionArn,
    };
  }

  /**
   * Create or load certificate; for creation DNS validation in hosted zone is used
   */
  private initCertificate(props: RouterProps): acm.ICertificate | undefined {
    return loadCertificate(
      this,
      props.certificate,
      () =>
        new acm.DnsValidatedCertificate(this, 'Certificate', {
          domainName: `*.${props.publicDomain}`,
          hostedZone: props.hostedZone,
          //validation: acm.CertificateValidation.fromDns(this.hostedZone),
        })
    );
  }

  /**
   * Either build docker image on the fly and upload to ECR of this CDK; or use provided
   * @param props
   */
  private initImage(props: RouterProps): ecs.ContainerImage {
    if (typeof props.image !== 'string' && props.image) {
      return props.image;
    }
    const { image } = service.loadContainerImage(
      this,
      props.image ?? defaultRouterImage,
      props.imageVersion ?? defaultRouterImageVersion
    );
    return image;
  }

  /**
   * Creates ECR task that runs the Traefik container image
   * @param props
   * @param image
   */
  private createService(props: RouterProps, image: ecs.ContainerImage): ecs.FargateService {
    // create task definition
    const taskDefinition = new ecs.FargateTaskDefinition(this, 'TaskDefinition', {
      cpu: this.resources.cpu,
      memoryLimitMiB: this.resources.memory,
      taskRole: this.grantPrincipal,
    });

    const logging = util.ecsLogDriver(props.logging ?? '/cluster/router');

    const container = taskDefinition.addContainer('Container', {
      logging,
      /* dockerLabels: {
        'traefik.domain': props.domainName,
      }, */
      environment: {
        TRAEFIK_ENTRYPOINTS_HTTP: 'true',
        TRAEFIK_ENTRYPOINTS_HTTP_ADDRESS: ':80',
        TRAEFIK_LOG: 'true',
        TRAEFIK_LOG_LEVEL: props.logLevel ?? 'ERROR',
        TRAEFIK_PING: 'true',
        TRAEFIK_PING_ENTRYPOINT: 'http',
        TRAEFIK_PROVIDERS_ECS: 'true',
        TRAEFIK_PROVIDERS_ECS_REFRESHSECONDS: '10',
        TRAEFIK_PROVIDERS_ECS_CLUSTERS: props.cluster.clusterName,
        TRAEFIK_PROVIDERS_ECS_REGION: cdk.Aws.REGION,
      },
      image,
    });

    container.addPortMappings({
      containerPort: 80,
      hostPort: 80,
      protocol: ecs.Protocol.TCP,
    });

    let args: Partial<ecs.FargateServiceProps> = {};
    if (this.resources.scaling.mode === 'fixed') {
      args = { ...args, desiredCount: this.resources.scaling.amount };
    }

    const service = new ecs.FargateService(this, 'Service', {
      cluster: props.cluster,
      taskDefinition,
      ...args,
    });

    return service;
  }

  /**
   * Configures services scaling configuration
   */
  private setupAutoScaling(): void {
    new service.Scaler(this, 'Scaler', {
      scaling: this.resources.scaling,
      service: this.service,
    });
  }

  /**
   * Creates the ALB that exposes the ECS task to the interwebs
   * @param props
   */
  private initLoadBalancer(props: RouterProps): elbv2.ApplicationLoadBalancer {
    const loadBalancer = new elbv2.ApplicationLoadBalancer(this, 'LoadBalancer', {
      vpc: props.cluster.vpc,
      http2Enabled: true,
      internetFacing: true,
      // TODO: idle timeout etc
    });

    // create HTTP listener and init list of forwarders (whose default action will be to forward to target(s))
    const createHttpListener = (action?: elbv2.ListenerAction): elbv2.ApplicationListener =>
      loadBalancer.addListener('HttpListener', {
        port: 80,
        open: true,
        defaultAction: action,
      });
    const forwarders: elbv2.ApplicationListener[] = [];

    // with certificate; create HTTPS listener
    if (this.certificate) {
      forwarders.push(
        loadBalancer.addListener('HttpsListener', {
          port: 443,
          open: true,
          certificates: [this.certificate],
        })
      );

      // redirect all HTTP requests to HTTPS? (default)
      if (props.enforceHttps ?? defaultEnforceHttps) {
        createHttpListener(
          elbv2.ListenerAction.redirect({
            port: '443',
            protocol: 'HTTPS',
            permanent: true,
          })
        );
      } else {
        forwarders.push(createHttpListener());
      }
    } else {
      forwarders.push(createHttpListener());
    }

    // optional: safe-guard load balancer access by checking for
    //   user defined request header and value. Use case: behind CloudFront
    const conditions = props.requireHeader
      ? {
          conditions: [elbv2.ListenerCondition.httpHeader(props.requireHeader.name, props.requireHeader.values)],
          priority: 1,
        }
      : {};

    // route incoming traffic for listeners
    forwarders.forEach((listener) => {
      listener.addTargets('EcsService', {
        port: 80,
        targets: [
          this.service.loadBalancerTarget({
            containerName: 'Container',
            containerPort: 80,
          }),
        ],
        healthCheck: {
          path: '/ping',
        },
        ...conditions,
      });
      if ('priority' in conditions) {
        listener.addAction('MissingHeader', {
          action: elbv2.ListenerAction.fixedResponse(403, {
            contentType: 'text/plain',
            messageBody: 'Forbidden',
          }),
        });
      }
    });

    return loadBalancer;
  }

  /**
   * Routes all subdomains (wildcard) of public domain to load balancer
   *
   * <code>
   *   foo-app.public.domain \                     / service foo-app
   *    barbar.public.domain  --> load balancer --> service barbar
   *         *.public.domain /                     \ etc
   * </code>
   *
   * @param props
   */
  private routeToLoadBalancer(props: RouterProps): void {
    [`*.${props.publicDomain}`].forEach((domain) => {
      new route53.ARecord(this, `RouterRecord${domain}`, {
        recordName: domain,
        zone: props.hostedZone,
        target: route53.RecordTarget.fromAlias(new targets.LoadBalancerTarget(this.loadBalancer)),
      });
    });
  }

  private publishInClusterNamespace(props: RouterProps) {
    const service = new servicediscovery.Service(this, 'NamespaceRecord', {
      namespace: props.cluster!.defaultCloudMapNamespace!,
      name: '_router',
      description: 'Traefik router',
    });
    this.service.associateCloudMapService({ service });
  }

  /**
   * Creates IAM role, including policy, to be attached to ECS task, so that
   * Treafik is allowed to scan the ECS cluster it runs in
   * @param props
   */
  private createServiceTaskRole(props: RouterProps) {
    // TODO: specify conditions for IAM principal, so that only they can..
    /* const conditions = {
      conditions: {
        "StringEquals": {
          "aws:sourceVpc": props.cluster.vpc.vpcId,
        }
      }
    }; */
    return new iam.Role(this, 'RouterTaskRole', {
      description:
        'Role that router service container assumes so that it communicate with ECS and EC2 APIs to auto-detect services (containers) to route traffic to',
      assumedBy: new iam.ServicePrincipal('ecs-tasks.amazonaws.com'),
      inlinePolicies: {
        TraefikTaskRolePolicy: new iam.PolicyDocument({
          statements: [
            // policy for traefik from: https://doc.traefik.io/traefik/providers/ecs/#policy
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              resources: ['*'],
              actions: [
                'ec2:DescribeInstances',
                'ecs:ListClusters',
                'ecs:DescribeClusters',
                'ecs:ListTasks',
                'ecs:DescribeTasks',
                'ecs:DescribeContainerInstances',
                'ecs:DescribeTaskDefinition',
              ],
            }),
          ],
        }),
      },
    });
  }
}
