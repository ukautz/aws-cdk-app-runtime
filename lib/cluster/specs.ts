import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as route53 from '@aws-cdk/aws-route53';
import * as servicediscovery from '@aws-cdk/aws-servicediscovery';
import { keys as compiledKeysOf } from 'ts-transformer-keys';
import { SpecUtil } from '../util/spec';
import { RouterSpecs } from './router';

/**
 * Specs (State) that the runtime cluster exports, that will be required by other
 * runtime components within the runtime cluster, like runtime services.
 */
export interface Specs extends RouterSpecs {
  /**
   * Name of the CloudMap namespace, i.e. the cluster internal hostname
   */
  privateDomain: string;

  /**
   * The full public domain names, like 'apps.domain.tld', of which sub-domains like '<my-app>.apps.domain.tld>', will be routing traffic to service <my-app>
   */
  publicDomain: string;

  /**
   * ARN of the hosted zone that holds the apex and wildcard records pointing to the load balancer
   */
  zoneArn: string;

  /**
   * ID of the hosted zone that holds the apex and wildcard records pointing to the load balancer
   */
  zoneId: string;

  /**
   * Name of the hosted zone that holds the apex and wildcard records pointing to the load balancer
   */
  zoneName: string;

  /**
   * Comma separated list of availability zones this cluster is using
   */
  availabilityZones: string;

  /**
   * ARN of the ECS cluster
   */
  clusterArn: string;

  /**
   * Name of the ECS cluster
   */
  clusterName: string;

  /**
   * ARN of the cloudmap namespace for internal service name resolution
   */
  namespaceArn: string;

  /**
   * ID of the cloudmap namespace for internal service name resolution
   */
  namespaceId: string;

  /**
   * Name of the cloudmap namespace for internal service name resolution
   */
  namespaceName: string;

  /**
   * ID of the VPC the ECS cluster runs in
   */
  vpcId: string;
}

/**
 * Names of keys in the Runtime Cluster Specification
 */
export const specKeys = compiledKeysOf<Specs>().sort();

export class Specs implements Specs {
  static make = (generator?: (prop: string) => string): Specs => SpecUtil.make<Specs>(specKeys, generator);
  static fromContext = (scope: cdk.Construct, prefix?: string): Specs => SpecUtil.fromContext(specKeys, scope, prefix);
  static fromSsm = (scope: cdk.Construct, prefix?: string): Specs => SpecUtil.fromSsm<Specs>(specKeys, scope, prefix);
  static toSsm = (scope: cdk.Construct, prefix: string, specs: Specs, secure?: boolean): void =>
    SpecUtil.toSsm<Specs>(scope, prefix, specs, secure);

  /**
   * Loads VPC defined in cluster specs
   */
  static lookupVpc = (scope: cdk.Construct, id: string, specs: Specs): ec2.IVpc =>
    ec2.Vpc.fromLookup(scope, id, {
      vpcId: specs.vpcId,
    });

  /**
   * Loads ECS cluster defined in cluster specs
   */
  static lookupCluster = (scope: cdk.Construct, id: string, specs: Specs, vpc?: ec2.IVpc): ecs.ICluster =>
    ecs.Cluster.fromClusterAttributes(scope, id, {
      clusterName: specs.clusterName,
      securityGroups: [],
      vpc: vpc ?? Specs.lookupVpc(scope, `${id}Vpc`, specs),
    });

  /**
   * Loads cluster cloudmap / service discovery namespace (for private domain)
   */
  static lookupPrivateNamespace = (
    scope: cdk.Construct,
    id: string,
    specs: Specs
  ): servicediscovery.IPrivateDnsNamespace =>
    servicediscovery.PrivateDnsNamespace.fromPrivateDnsNamespaceAttributes(scope, id, specs);
}
