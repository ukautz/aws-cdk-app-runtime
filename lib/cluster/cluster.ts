import {
  aws_certificatemanager as acm,
  aws_ec2 as ec2,
  aws_ecs as ecs,
  aws_route53 as route53,
  aws_servicediscovery as servicediscovery,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { loadHostedZone } from '../util/loaders';
import { Router, RouterPropsExposed } from './router';
import { Specs } from './specs';

export interface ClusterProps {
  /**
   * Name of the CloudMap namespace, i.e. the cluster internal hostname
   */
  readonly privateDomain: string;

  /**
   * The full public domain names, like 'apps.domain.tld', of which sub-domains like '<my-app>.apps.domain.tld>', will be routing traffic to service <my-app>
   */
  readonly publicDomain: string;

  /**
   * opt: existing hosted zones for used domains (hosted zone will be created if not given)
   */
  readonly hostedZone?: string | route53.IHostedZone;

  /**
   * Determines whether and how certificate is used:
   * * `false`: no certificate, route only port 80
   * * `true`:  create a certificate, assuming parentHostedZone is provided, route 80 to 443
   * * string: ARN to certificate, route 80 to 443
   * * a certificate
   */
  readonly certificate?: boolean | string | acm.ICertificate;

  /**
   * Optional ID of existing VPC. If not provided new VPC will be created
   */
  readonly vpc?: string | ec2.IVpc;

  /**
   * Optional max amount of AZ to span the cluster one. Assumes vpcId attribute is not defined
   */
  readonly maxAzs?: number;

  /**
   * Optional NAT instance size. If not provided then nat gateway is being used. Assumes vpcId attribute is not defined
   */
  readonly natInstanceType?: string;

  /**
   * Router specification, or default
   */
  readonly routerProps?: RouterPropsExposed;
}

/**
 * Cluster provides the Runtime part of Application Runtime.
 *
 * Components:
 * - Public domain, under which services will be exposed
 * - Private domain, under which inter-service traffic should flow
 * - A VPC, that can be provided or managed
 * - A managed ECS cluster (including a CloudMap of the private domain)
 * - A hosted zone for the public domain
 *
 * Considerations:
 * - Like to see in EKS. Option or Class? What is shared?
 */
export class Cluster extends Construct {
  public readonly privateDomain: string;
  public readonly publicDomain: string;
  public readonly hostedZone: route53.IHostedZone;
  public readonly vpc: ec2.IVpc;
  public readonly cluster: ecs.ICluster;
  public readonly router: Router;

  constructor(scope: Construct, id: string, props: ClusterProps) {
    super(scope, id);

    this.privateDomain = props.privateDomain;
    this.publicDomain = props.publicDomain;

    this.hostedZone = loadHostedZone(this, props.hostedZone, () =>
      route53.HostedZone.fromLookup(this, 'HostedZone', {
        domainName: this.publicDomain,
      })
    );

    if (props.vpc && (props.natInstanceType || props.maxAzs)) {
      throw new Error('natInstanceType and maxAzs not allowed when vpcId is provided');
    }

    this.vpc = this.initVpc(props);
    this.cluster = this.initCluster(props);
    this.router = new Router(this, 'Router', {
      ...props.routerProps,
      publicDomain: props.publicDomain,
      certificate: props.certificate,
      cluster: this.cluster,
      hostedZone: this.hostedZone,
    });
  }

  public get specs(): Specs {
    return {
      ...this.router.specs,
      availabilityZones: this.vpc.availabilityZones.join(','),
      clusterArn: this.cluster.clusterArn,
      clusterName: this.cluster.clusterName,
      namespaceArn: this.cluster.defaultCloudMapNamespace!.namespaceArn,
      namespaceId: this.cluster.defaultCloudMapNamespace!.namespaceId,
      namespaceName: this.cluster.defaultCloudMapNamespace!.namespaceName,
      privateDomain: this.privateDomain,
      publicDomain: this.publicDomain,
      vpcId: this.vpc.vpcId,
      zoneArn: this.hostedZone.hostedZoneArn,
      zoneId: this.hostedZone.hostedZoneId,
      zoneName: this.hostedZone.zoneName,
    } as Specs;
  }

  /**
   * Create or load the VPC
   * @param props
   */
  private initVpc(props: ClusterProps): ec2.IVpc {
    if (typeof props.vpc === 'string') {
      return ec2.Vpc.fromLookup(this, 'Vpc', { vpcId: props.vpc });
    } else if (props.vpc) {
      return props.vpc;
    }
    return new ec2.Vpc(this, 'Vpc', {
      maxAzs: props.maxAzs,
      subnetConfiguration: ec2.Vpc.DEFAULT_SUBNETS,
      ...(props.natInstanceType
        ? {
            natGatewayProvider: new ec2.NatInstanceProvider({
              instanceType: new ec2.InstanceType(props.natInstanceType),
            }),
          }
        : {}),
    });
  }

  /**
   * Create the ECS Cluster all tasks will run in
   * @param props
   */
  private initCluster(props: ClusterProps): ecs.ICluster {
    return new ecs.Cluster(this, 'Cluster', {
      vpc: this.vpc,
      defaultCloudMapNamespace: {
        name: props.privateDomain,
        type: servicediscovery.NamespaceType.DNS_PRIVATE,
      },
    });
  }
}
