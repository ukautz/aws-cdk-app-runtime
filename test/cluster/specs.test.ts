import '@aws-cdk/assert/jest';
import { Specs as ClusterSpecs, specKeys as clusterSpecKeys } from '../../lib/cluster/specs';
import { assertSpecs } from '../util/specs.suite';

describe('Specs', () => {
  test('Can list keys', () => {
    expect(clusterSpecKeys).toEqual(
      [
        'availabilityZones',
        'clusterArn',
        'clusterName',
        'namespaceArn',
        'namespaceId',
        'namespaceName',
        'privateDomain',
        'publicDomain',
        'routerArn',
        'routerCertificateArn',
        'routerLoadBalancerArn',
        'routerLoadBalancerDnsName',
        'routerLoadBalancerSecurityGroupId',
        'routerRoleArn',
        'routerSecurityGroupIds',
        'routerTaskDefinitionArn',
        'vpcId',
        'zoneArn',
        'zoneId',
        'zoneName',
      ].sort()
    );
  });
  assertSpecs<ClusterSpecs>({
    name: 'Cluster',
    keys: clusterSpecKeys,
    make: ClusterSpecs.make,
    fromContext: ClusterSpecs.fromContext,
    fromSsm: ClusterSpecs.fromSsm,
    toSsm: ClusterSpecs.toSsm,
  });
});
