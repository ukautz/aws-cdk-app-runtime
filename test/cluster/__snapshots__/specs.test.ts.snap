// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Specs Cluster Store Encrypted to Ssm Matches Snapshot 1`] = `
Object {
  "Resources": Object {
    "AvailabilityZonesSpec9E4B3187": Object {
      "Properties": Object {
        "Name": "/prefixed/availabilityZones",
        "Type": "SecureString",
        "Value": "availabilityZones",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ClusterArnSpec575C9810": Object {
      "Properties": Object {
        "Name": "/prefixed/clusterArn",
        "Type": "SecureString",
        "Value": "clusterArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ClusterNameSpecE20605BB": Object {
      "Properties": Object {
        "Name": "/prefixed/clusterName",
        "Type": "SecureString",
        "Value": "clusterName",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "NamespaceArnSpec9EB19FED": Object {
      "Properties": Object {
        "Name": "/prefixed/namespaceArn",
        "Type": "SecureString",
        "Value": "namespaceArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "NamespaceIdSpecBFA4A494": Object {
      "Properties": Object {
        "Name": "/prefixed/namespaceId",
        "Type": "SecureString",
        "Value": "namespaceId",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "NamespaceNameSpecB3BB8809": Object {
      "Properties": Object {
        "Name": "/prefixed/namespaceName",
        "Type": "SecureString",
        "Value": "namespaceName",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "PrivateDomainSpec895E9A72": Object {
      "Properties": Object {
        "Name": "/prefixed/privateDomain",
        "Type": "SecureString",
        "Value": "privateDomain",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "PublicDomainSpecD88DC9DD": Object {
      "Properties": Object {
        "Name": "/prefixed/publicDomain",
        "Type": "SecureString",
        "Value": "publicDomain",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterArnSpecA6735D98": Object {
      "Properties": Object {
        "Name": "/prefixed/routerArn",
        "Type": "SecureString",
        "Value": "routerArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterCertificateArnSpec7CF705EB": Object {
      "Properties": Object {
        "Name": "/prefixed/routerCertificateArn",
        "Type": "SecureString",
        "Value": "routerCertificateArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterLoadBalancerArnSpec7857920B": Object {
      "Properties": Object {
        "Name": "/prefixed/routerLoadBalancerArn",
        "Type": "SecureString",
        "Value": "routerLoadBalancerArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterLoadBalancerDnsNameSpec38B41CDB": Object {
      "Properties": Object {
        "Name": "/prefixed/routerLoadBalancerDnsName",
        "Type": "SecureString",
        "Value": "routerLoadBalancerDnsName",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterLoadBalancerSecurityGroupIdSpec86735FD5": Object {
      "Properties": Object {
        "Name": "/prefixed/routerLoadBalancerSecurityGroupId",
        "Type": "SecureString",
        "Value": "routerLoadBalancerSecurityGroupId",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterRoleArnSpec568B75F8": Object {
      "Properties": Object {
        "Name": "/prefixed/routerRoleArn",
        "Type": "SecureString",
        "Value": "routerRoleArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterSecurityGroupIdsSpec721726F2": Object {
      "Properties": Object {
        "Name": "/prefixed/routerSecurityGroupIds",
        "Type": "SecureString",
        "Value": "routerSecurityGroupIds",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterTaskDefinitionArnSpec129A4FBC": Object {
      "Properties": Object {
        "Name": "/prefixed/routerTaskDefinitionArn",
        "Type": "SecureString",
        "Value": "routerTaskDefinitionArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "VpcIdSpec0C86AA6B": Object {
      "Properties": Object {
        "Name": "/prefixed/vpcId",
        "Type": "SecureString",
        "Value": "vpcId",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ZoneArnSpec22EE4B29": Object {
      "Properties": Object {
        "Name": "/prefixed/zoneArn",
        "Type": "SecureString",
        "Value": "zoneArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ZoneIdSpecB68C855E": Object {
      "Properties": Object {
        "Name": "/prefixed/zoneId",
        "Type": "SecureString",
        "Value": "zoneId",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ZoneNameSpec724EA8E0": Object {
      "Properties": Object {
        "Name": "/prefixed/zoneName",
        "Type": "SecureString",
        "Value": "zoneName",
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;

exports[`Specs Cluster Store to Ssm Matches Snapshot 1`] = `
Object {
  "Resources": Object {
    "AvailabilityZonesSpec9E4B3187": Object {
      "Properties": Object {
        "Name": "/prefixed/availabilityZones",
        "Type": "String",
        "Value": "availabilityZones",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ClusterArnSpec575C9810": Object {
      "Properties": Object {
        "Name": "/prefixed/clusterArn",
        "Type": "String",
        "Value": "clusterArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ClusterNameSpecE20605BB": Object {
      "Properties": Object {
        "Name": "/prefixed/clusterName",
        "Type": "String",
        "Value": "clusterName",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "NamespaceArnSpec9EB19FED": Object {
      "Properties": Object {
        "Name": "/prefixed/namespaceArn",
        "Type": "String",
        "Value": "namespaceArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "NamespaceIdSpecBFA4A494": Object {
      "Properties": Object {
        "Name": "/prefixed/namespaceId",
        "Type": "String",
        "Value": "namespaceId",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "NamespaceNameSpecB3BB8809": Object {
      "Properties": Object {
        "Name": "/prefixed/namespaceName",
        "Type": "String",
        "Value": "namespaceName",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "PrivateDomainSpec895E9A72": Object {
      "Properties": Object {
        "Name": "/prefixed/privateDomain",
        "Type": "String",
        "Value": "privateDomain",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "PublicDomainSpecD88DC9DD": Object {
      "Properties": Object {
        "Name": "/prefixed/publicDomain",
        "Type": "String",
        "Value": "publicDomain",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterArnSpecA6735D98": Object {
      "Properties": Object {
        "Name": "/prefixed/routerArn",
        "Type": "String",
        "Value": "routerArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterCertificateArnSpec7CF705EB": Object {
      "Properties": Object {
        "Name": "/prefixed/routerCertificateArn",
        "Type": "String",
        "Value": "routerCertificateArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterLoadBalancerArnSpec7857920B": Object {
      "Properties": Object {
        "Name": "/prefixed/routerLoadBalancerArn",
        "Type": "String",
        "Value": "routerLoadBalancerArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterLoadBalancerDnsNameSpec38B41CDB": Object {
      "Properties": Object {
        "Name": "/prefixed/routerLoadBalancerDnsName",
        "Type": "String",
        "Value": "routerLoadBalancerDnsName",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterLoadBalancerSecurityGroupIdSpec86735FD5": Object {
      "Properties": Object {
        "Name": "/prefixed/routerLoadBalancerSecurityGroupId",
        "Type": "String",
        "Value": "routerLoadBalancerSecurityGroupId",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterRoleArnSpec568B75F8": Object {
      "Properties": Object {
        "Name": "/prefixed/routerRoleArn",
        "Type": "String",
        "Value": "routerRoleArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterSecurityGroupIdsSpec721726F2": Object {
      "Properties": Object {
        "Name": "/prefixed/routerSecurityGroupIds",
        "Type": "String",
        "Value": "routerSecurityGroupIds",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "RouterTaskDefinitionArnSpec129A4FBC": Object {
      "Properties": Object {
        "Name": "/prefixed/routerTaskDefinitionArn",
        "Type": "String",
        "Value": "routerTaskDefinitionArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "VpcIdSpec0C86AA6B": Object {
      "Properties": Object {
        "Name": "/prefixed/vpcId",
        "Type": "String",
        "Value": "vpcId",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ZoneArnSpec22EE4B29": Object {
      "Properties": Object {
        "Name": "/prefixed/zoneArn",
        "Type": "String",
        "Value": "zoneArn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ZoneIdSpecB68C855E": Object {
      "Properties": Object {
        "Name": "/prefixed/zoneId",
        "Type": "String",
        "Value": "zoneId",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ZoneNameSpec724EA8E0": Object {
      "Properties": Object {
        "Name": "/prefixed/zoneName",
        "Type": "String",
        "Value": "zoneName",
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;
